src/core/textlayout.js

```diff
@@ -0,0 +1,176 @@
+      this.checked = 0;
+    }
+  };
+  var deferred = Promise.resolve();
+
+  TextLayoutEvaluator.prototype = {
+    overlapBy : function (min1, min2, max1, max2, by) {
+        var d = Math.min(max1,max2) - Math.max(min1,min2);
+        return d > (max1-min1) * by || d > (max2-min2) * by;
+    },
+    
+    addToQuadTree: function(o, i, styles) {
+      var style = styles[o.fontName];
+      if (isAllWhitespace(o.str)) {
+        // Whitespace elements aren't visible, but they're used for copy/paste.
+        o.isWhitespace = true;
+      } else {
+        // Line numbers can mess up flow, so we detect and handle them them.
+        var n = Number(o.str.replace(/^\s+|\s+$/g,''));
+        if(n % 1 === 0) {
+          o.asInt = n;
+        }
+      }
+      var tx = o.transform;
+      
+      var angle = Math.atan2(tx[1], tx[0]);
+      if (style.vertical) {
+        angle += Math.PI / 2;
+      }
+      var fontHeight = Math.sqrt((tx[2] * tx[2]) + (tx[3] * tx[3]));
+      var fontAscent = fontHeight;
+      if (style.ascent) {
+        fontAscent = style.ascent * fontAscent;
+      } else if (style.descent) {
+        fontAscent = (1 + style.descent) * fontAscent;
+      }
+
+      // Set y,x to the bottom left. They're the smaller values.
+      o.y = tx[5];
+      o.x = (angle === 0) ? tx[4] : (tx[4] + (fontAscent * Math.sin(angle)));
+      o.vertical = style.vertical;
+      
+      o.id = i;         // Used to uniquely identify object.
+      o.right = null;   // The nearest object to the right.
+      o.bottom = null;  // The nearest object to the left.
+          
+      this.quadtree.insert(o);
+    },
+    
+    calculateTextFlow: function (bounds, objs, styles) {
+      var timeSlotManager = new TimeSlotManager();
+      var self = this;
+      self.bounds = bounds;
+      
+      // Put everything into the quadtree for O(logN) positional look ups.
+      self.quadtree = new QuadTree(bounds, 4, 16);
+      for (var i = 0, len = objs.length; i < len; i++) {
+        self.addToQuadTree(objs[i], i, styles);
+      }
+      
+      // The top item has a big impact on the flow, so track it.
+      var top_obj = null;
+      
+      // Set each element's padding to run to the nearest right and bottom 
+      // element. The padding ensures that text selection works.
+      for (i = 0; i < len; i++) {
+        var d = objs[i];
+        
+        // Track the top item. Not used yet, but can help reorder layout flow.
+        if(!d.isWhitespace && (top_obj === null || d.y > top_obj.y)) {
+            top_obj = d;
+        }
+        
+        var dn, it;
+        // Find the first object to the right.
+        it = self.quadtree.retrieve_xinc(d.x+d.width,d.y,d.height);
+        while(dn = it.next()) {
+          if(dn.id !== d.id) {
+            d.right = dn.id;
+            break;
+          }
+        }
+        // Find the left.
+        it = self.quadtree.retrieve_xdec(d.x,d.y,d.height);
+        while(dn = it.next()) {
```

Comment:
Nit: add a space before `(`. This applies to more `if` and `while` loops in this file, so please check all of them to make sure that the code adheres to the code style guide.